components/camera/rtsp/rtsp.go

```diff
@@ -0,0 +1,168 @@
+
+import (
+	"bytes"
+	"context"
+	"image"
+	"image/jpeg"
+	"sync/atomic"
+
+	"github.com/aler9/gortsplib/v2"
+	"github.com/aler9/gortsplib/v2/pkg/format"
+	"github.com/aler9/gortsplib/v2/pkg/liberrors"
+	"github.com/aler9/gortsplib/v2/pkg/url"
+	"github.com/edaniels/golog"
+	"github.com/edaniels/gostream"
+	"github.com/pion/rtp"
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+
+	"go.viam.com/rdk/components/camera"
+	"go.viam.com/rdk/config"
+	"go.viam.com/rdk/registry"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/rimage/transform"
+	"go.viam.com/rdk/utils"
+)
+
+var model = resource.NewDefaultModel("rtsp")
+
+func init() {
+	registry.RegisterComponent(camera.Subtype, model, registry.Component{
+		Constructor: func(ctx context.Context, _ registry.Dependencies, cfg config.Component, logger golog.Logger) (interface{}, error) {
+			attrs, ok := cfg.ConvertedAttributes.(*Attrs)
+			if !ok {
+				return nil, utils.NewUnexpectedTypeError(attrs, cfg.ConvertedAttributes)
+			}
+			return NewRTSPCamera(ctx, attrs, logger)
+		},
+	})
+
+	config.RegisterComponentAttributeMapConverter(
+		camera.Subtype,
+		model,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var conf Attrs
+			attrs, err := config.TransformAttributeMapToStruct(&conf, attributes)
+			if err != nil {
+				return nil, err
+			}
+			result, ok := attrs.(*Attrs)
+			if !ok {
+				return nil, utils.NewUnexpectedTypeError(result, attrs)
+			}
+			return result, nil
+		},
+		&Attrs{},
+	)
+}
+
+// Attrs are the config attributes for an RTSP camera model.
+type Attrs struct {
+	Address          string                             `json:"rtsp_address"`
+	IntrinsicParams  *transform.PinholeCameraIntrinsics `json:"intrinsic_parameters,omitempty"`
+	DistortionParams *transform.BrownConrady            `json:"distortion_parameters,omitempty"`
+}
+
+// rtspCamera contains the rtsp client, and the reader function that fulfills the camera interface.
+type rtspCamera struct {
+	gostream.VideoReader
+	client     *gortsplib.Client
+	cancelFunc context.CancelFunc
+}
+
+// Close closes the camera.
+func (rc *rtspCamera) Close(ctx context.Context) error {
+	clientTerminated := liberrors.ErrClientTerminated{}
+	rc.cancelFunc()
+	err := rc.client.Close()
+	if err != nil && !errors.Is(err, clientTerminated) {
+		return err
+	}
+	return nil
+}
+
+// NewRTSPCamera creates a camera client for an RTSP given given the server URL.
+// Right now, only supports servers that have MJPEG video tracks.
+func NewRTSPCamera(ctx context.Context, attrs *Attrs, logger golog.Logger) (camera.Camera, error) {
+	c := &gortsplib.Client{}
+	// parse URL
+	u, err := url.Parse(attrs.Address)
+	if err != nil {
+		return nil, err
+	}
+	// connect to the server - be sure to close it if setup fails.
+	err = c.Start(u.Scheme, u.Host)
+	if err != nil {
+		return nil, err
+	}
+	// find published tracks
+	tracks, baseURL, _, err := c.Describe(u)
+	if err != nil {
+		return nil, multierr.Combine(err, c.Close())
```

Comment:
I think I will skip putting it in a defer, because I don't want the client to Close() at the end of the NewRTSPCamera function if everything went well once it returns. Setting up the client takes a few steps, and it can fail at any stage, and I want it to close only if it failed to pass any of its set up stages.

Reply:
Ah, I see now about the `var successful bool` way of doing this, will implement

Reply:
done